static/app/views/ddm/useGetChartPalette.tsx

```diff
@@ -0,0 +1,62 @@
+import {useCallback, useRef} from 'react';
+
+import theme from 'sentry/utils/theme';
+
+const CACHE_SIZE = 20; // number of palettes to cache
+
+// We do length - 2 to be aligned with the colors in other parts of the app (copy-pasta)
+// We use Math.max to avoid numbers < -1 as then `getColorPalette` returns undefined (not typesafe because of array access)
+export function getChartColors(seriesCount: number) {
+  return theme.charts.getColorPalette(Math.max(seriesCount - 2, -1));
+}
+
+export function getCachedChartPalette(
+  cache: Record<string, string>[],
+  seriesNames: string[]
+): (seriesName: string) => string {
+  // We do length - 2 to be aligned with the colors in other parts of the app (copy-pasta)
+  // We use Math.max to avoid numbers < -1 as then `getColorPalette` returns undefined (not typesafe because of array access)
+  const defaultColors = getChartColors(seriesNames.length);
+
+  // Check if we already have a palette that includes all of the given seriesNames
+  // We reverse the cache to get the most recent palettes first
+  const cachedPalette = [...cache]
+    .reverse()
+    .find(palette => seriesNames.every(seriesName => seriesName in palette));
```

Comment:
nit: this makes cache lookups On*m. May or may not be an issue depending on how many series we expect per entry, but unless it's a lot, it doesn't matter

Reply:
Yeah, it's likely fine. If you have a max of 20 entries over a max of 20 cache entries, that is still only 400 iterations. You have a few orders of magnitude to go before it becomes a problem :)